WebAuthn docs update (#25922)
mdn/content: files/en-us/web/api/web_authentication_api/webauthn_extensions/index.md

  - api.CredentialsContainer.get.publicKey_option.extensions
---

{{DefaultAPISidebar("Web Authentication API")}}

The [Web Authentication API](/en-US/docs/Web/API/Web_Authentication_API) has a system of extensions — extra functionality that can be requested during credential creation ({{domxref("CredentialsContainer.create()","navigator.credentials.create()")}}) or authentication ({{domxref("CredentialsContainer.get()","navigator.credentials.get()")}}) operations. This article explains how to request WebAuthn extensions, how to retrieve information about the responses from those requests, and the extensions that are available — including browser support and expected inputs and outputs.

## How to use WebAuthn extensions

When invoking {{domxref("CredentialsContainer.create()","navigator.credentials.create()")}} or {{domxref("CredentialsContainer.get()","navigator.credentials.get()")}}, the `publicKey` object parameter required to initiate a WebAuthn flow can include an `extensions` property. The value of `extensions` is itself an object, the properties of which are the input values for any extensions the relying party wishes to request the use of in the method you are calling.

> Behind the scenes, request extensions receive the inputs, and are processed by the user agent (client) and/or the authenticator.

Author: I think I took something you said about request and response extensions and tried to run with it, but it didn't turn out very well. I've updated the above with:

> Behind the scenes, the inputs are processed by the user agent and/or the authenticator.

Does that work better?

For example, in a `publicKey` object for a `create()` call, we might want to request the use of two extensions:

> 1. The `credProps` extension, which allows relying parties to specify that a resident key is required, meaning that the credential is a client-side discoverable credential. This requires `extensions.credProps` to be set to `true`. This also requires `authenticatorSelection.requireResidentKey` to be set to `true` (visit [`create()` > `publicKey` object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure) and search for `requireResidentKey`). `credProps` is processed by the user agent only.

Author: Thanks for the info; I've updated the details.

}
```

We can then pass the `publicKey` object into a `create()` call to initiate the credential creation flow:

```js
navigator.credentials.create({ publicKey });
```

## Retrieving extension request results

If successful, the `create()` call will return a {{jsxref("Promise")}} that resolves with a {{domxref("PublicKeyCredential")}} object. Behind the scenes, once extension processing has completed, output extensions are triggered that communicate the results of the processing (although not in all cases — it is possible for extensions to have no output).

```js
navigator.credentials
  .create({ publicKey })
  .then((publicKeyCred) => {
    const myClientExtResults = publicKeyCred.getClientExtensionResults();
    // myClientExtResults will contain the output of processing
    // the "credProps" extension

    const authData = publicKeyCred.response.getAuthenticatorData();
    // authData will contain authenticator data, which will include
    // authenticator extension processing results, i.e. minPinLength
  })
  .catch((err) => {
    console.error(err);
  });
```

As demonstrated by the above code snippet, there are two different places to find your output extension results:

> 1. You can find the results of client (user agent) extension processing by calling the {{domxref("PublicKeyCredential.getClientExtensionResults()")}} method. This returns a {{jsxref("Map", "map")}}, with each entry being an extensions' identifier string as the key, and the output from the processing of the extension by the client as the value. In the example above, if the browser supported the `credProps` extension and it was processed correctly, the `myClientExtResults` map object would contain one entry, `"credProps"`, with a value of `{ rk: true }`. This would verify that the created credential does indeed require a resident key when used for authentication.

Author: I've updated this bit too.

   Authenticator data takes the form of an {{jsxref("ArrayBuffer")}} with a consistent structure — see [authenticator data](/en-US/docs/Web/API/Web_Authentication_API/Authenticator_data). The authenticator extension results data is always found in a section at the end, as a [CBOR map](https://cbor.io/) representing the results. See {{domxref("AuthenticatorAssertionResponse.authenticatorData")}} for a detailed description of the complete authenticator data structure.

   Back to our example, if the relying party is authorized to receive the `minPinLength` value, the authenticator data would contain a representation of it in the following form: `"minPinLength": uint`.

## Available extensions

The extensions below do not represent an exhaustive list of all the extensions available. We have elected to document extensions that we know to be standardized and supported by at least one rendering engine.

### `appid`

- Usable in: Authentication ({{domxref("CredentialsContainer.get()","get()")}})
- Processed by: User agent
- Specification: [FIDO AppID Extension (appid)](https://w3c.github.io/webauthn/#sctn-appid-extension)

> Allows a relying party to request an assertion for a credential previously registered using the legacy FIDO U2F JavaScript API, avoiding the hassle of re-registering the credential. The `appid` is that API's equivalent to the `rpId` in WebAuthn.

Author: I've added this info; thanks.

#### Input

The `publicKey`'s `extensions` property must contain an `appid` property, the value of which is the relying party identifier used in the legacy API. For example:

```js
extensions: {
  appid: "https://accounts.example.com";
}
```

> You must also list the FIDO U2F credentials in the `publicKey`'s `allowCredentials` property, for example:

Author: updated.

#### Input

The `publicKey`'s `extensions` property must contain an `appidExclude` property, the value of which is the identifier of the relying party requesting to exclude authenticators by legacy FIDO U2F credentials. For example:

```js
extensions: {
  appidExclude: "https://accounts.example.com";
}
```

You can then list FIDO U2F credentials in the `publicKey`'s `excludeCredentials` property, for example:

```js
excludeCredentials: {
  [
    id: arrayBuffer, // needs to contain decoded binary form of id
    transports: ["nfc", "usb"]
    type: "public-key"
  ]
}
```

#### Output

Outputs `appidExclude: true` if the extension was acted upon, or `appidExclude: false` otherwise.

### `credProps`

- Usable in: Registration ({{domxref("CredentialsContainer.create()","create()")}})
- Processed by: User agent
- Specification: [Credential Properties Extension (credProps)](https://w3c.github.io/webauthn/#sctn-authenticator-credential-properties-extension)

> Allows a relying party to specify credential properties; currently it only specifies that a resident key is required, meaning that the credential is a client-side discoverable credential (it may support specifying others in the future).

Author: Updated.

#### Input

The `publicKey`'s `extensions` property must contain a `credProps` property with a value of `true`:

```js
extensions: {
  credProps: true;
}
```

You must also set `authenticatorSelection.requireResidentKey` to `true`, which indicates that a resident key is required.

```js
authenticatorSelection: {
  requireResidentKey: true;
}

> ```

Author: Ah, OK. I got this confused. I've updated `authenticatorSelection` to contain `residentKey: "preferred"`. Is that correct?

If `rk` is set to `false` in the output, the credential is a server-side credential. If `rk` is not present in the output, then it is not known whether the credential is client-side discoverable or server-side.

### `credProtect`

- Usable in: Registration ({{domxref("CredentialsContainer.create()","create()")}})
- Processed by: Authenticator
- Specification: [Credential Protection (credProtect)](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#sctn-credProtect-extension)

Allows a relying party to specify a minimum credential protection policy when creating a credential.

#### Input

The `publicKey`'s `extensions` property must contain a `credentialProtectionPolicy` property specifying the protection level of the credential to be created, and a boolean `enforceCredentialProtectionPolicy` property specifying whether the `create()` call should fail rather than creating a credential that does not conform to the specified policy:

```js
extensions: {
  credentialProtectionPolicy: "userVerificationOptional",
  enforceCredentialProtectionPolicy: true
}
```

The available `credentialProtectionPolicy` values are as follows:

> - `"userVerificationOptional"`: User verification is optional. This is the default state if the extension is not specified. The equivalent `credProtect` value sent to the authenticator for processing is `0x01`.

Author: I've just copied the info off that page; it is pretty understandable.

---
title: Web Authentication extensions
slug: Web/API/Web_Authentication_API/WebAuthn_extensions
page-type: guide
browser-compat:
  - api.CredentialsContainer.create.publicKey_option.extensions
  - api.CredentialsContainer.get.publicKey_option.extensions
---

{{DefaultAPISidebar("Web Authentication API")}}

> The [Web Authentication API](/en-US/docs/Web/API/Web_Authentication_API) has a system of extensions — extra functionality that can be requested during credential creation ({{domxref("CredentialsContainer.create()","navigator.credentials.create()")}}) or authentication ({{domxref("CredentialsContainer.get()","navigator.credentials.get()")}}) operations. This article explains how to request WebAuthn extensions, how to retrieve information about the responses from those requests, and the extensions that are available — including browser support and expected inputs and outputs.

Reviewer: ```suggestion
The [Web Authentication API](/en-US/docs/Web/API/Web_Authentication_API) has a system of extensions — extra functionality that can be requested during credential creation ({{domxref("CredentialsContainer.create()", "navigator.credentials.create()")}}) or authentication ({{domxref("CredentialsContainer.get()", "navigator.credentials.get()")}}) operations. This article explains how to request WebAuthn extensions, retrieve information about the responses from those requests, and the available extensions — including browser support and expected inputs and outputs.
```

## How to use WebAuthn extensions

> When invoking {{domxref("CredentialsContainer.create()","navigator.credentials.create()")}} or {{domxref("CredentialsContainer.get()","navigator.credentials.get()")}}, the `publicKey` object parameter required to initiate a WebAuthn flow can include an `extensions` property. The value of `extensions` is itself an object, the properties of which are the input values for any extensions the relying party wishes to request the use of in the method you are calling.

Reviewer: ```suggestion
When invoking {{domxref("CredentialsContainer.create()", "navigator.credentials.create()")}} or {{domxref("CredentialsContainer.get()", "navigator.credentials.get()")}}, the `publicKey` object parameter required to initiate a WebAuthn flow can include an `extensions` property. The value of `extensions` is itself an object, the properties of which are the input values for any extensions the relying party wishes to request the use of in the method you call.
```

> Behind the scenes, the inputs are processed by the user agent and/or the authenticator.
> 
> For example, in a `publicKey` object for a `create()` call, we might want to request the use of two extensions:
> 
> 1. The `credProps` extension. Relying parties set `credProps` to request that the browser tells the relying party whether the credential is resident/discoverable after registration. This is useful when calling `create()` with `publicKey.authenticatorSelection.residentKey = "preferred"`. To request it, you also need to set `publicKey.extensions.credProps = true` when. The browser makes a credential and, depending on the type of authenticator used, it will be discoverable or not (for example, the FIDO2 authenticator would typically make it discoverable; FIDO1/U2F security key would be non-discoverable). `credProps` is processed by the user agent only.
> 2. The `minPinLength` extension, which allows relying parties to request the authenticator's minimum PIN length. This requires `extensions.minPinLength` to be set to `true`. `minPinLength` is processed by the authenticator, with the user agent only serving to pass the input data along to it.

Reviewer: ```suggestion
Behind the scenes, the inputs are processed by the user agent and/or the authenticator.

For example, in a `publicKey` object for a `create()` call, we might want to request the use of two extensions:

1. The `credProps` extension. Relying parties set `credProps` to request that the browser tells the relying party whether the credential is resident/discoverable after registration. This is useful when calling `create()` with `publicKey.authenticatorSelection.residentKey = "preferred"`. To request it, you also need to set `publicKey.extensions.credProps = true` when. The browser makes a credential and, depending on the type of authenticator used, it will be discoverable (for example, the FIDO2 authenticator would typically make it discoverable; FIDO1/U2F security key would be non-discoverable). `credProps` is processed by the user agent only.
2. The `minPinLength` extension allows relying parties to request the authenticator's minimum PIN length. This requires `extensions.minPinLength` to be set to `true`. `minPinLength` is processed by the authenticator, with the user agent only serving to pass the input data along to it.
```

    name: "jamiedoe",
    displayName: "Jamie Doe"
  },
  pubKeyCredParams: [ {type: "public-key", alg: -7} ],
  authenticatorSelection: {
    residentKey: "preferred"
  },
  extensions: {
    credProps: true,
    minPinLength: true
  }
}
```

We can then pass the `publicKey` object into a `create()` call to initiate the credential creation flow:

```js
navigator.credentials.create({ publicKey });
```

## Retrieving extension request results

If successful, the `create()` call will return a {{jsxref("Promise")}} that resolves with a {{domxref("PublicKeyCredential")}} object. Once extension processing has completed, the results of the processing are communicated in the response (although not in all cases — it is possible for extensions to have no output).

```js
navigator.credentials
  .create({ publicKey })
  .then((publicKeyCred) => {
    const myClientExtResults = publicKeyCred.getClientExtensionResults();
    // myClientExtResults will contain the output of processing
    // the "credProps" extension

    const authData = publicKeyCred.response.getAuthenticatorData();
    // authData will contain authenticator data, which will include

>     // authenticator extension processing results, i.e. minPinLength

Reviewer: ```suggestion
    // authenticator extension processing results, i.e., minPinLength
```

### `credProtect`

- Usable in: Registration ({{domxref("CredentialsContainer.create()","create()")}})
- Processed by: Authenticator
- Specification: [Credential Protection (credProtect)](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#sctn-credProtect-extension)

Allows a relying party to specify a minimum credential protection policy when creating a credential.

#### Input

The `publicKey`'s `extensions` property must contain a `credentialProtectionPolicy` property specifying the protection level of the credential to be created, and a boolean `enforceCredentialProtectionPolicy` property specifying whether the `create()` call should fail rather than creating a credential that does not conform to the specified policy:

```js
extensions: {
  credentialProtectionPolicy: "userVerificationOptional",
  enforceCredentialProtectionPolicy: true
}
```

The available `credentialProtectionPolicy` values are as follows:

- `"userVerificationOptional"`: User verification is optional. The equivalent `credProtect` value sent to the authenticator for processing is `0x01`.

> - `"userVerificationOptionalWithCredentialIDList"`: User verification is optional, only if the credential is discoverable (i.e. it is client-side discoverable). The equivalent `credProtect` value sent to the authenticator for processing is `0x02`.

Reviewer: ```suggestion
- `"userVerificationOptionalWithCredentialIDList"`: User verification is optional only if the credential is discoverable (i.e., it is client-side discoverable). The equivalent `credProtect` value sent to the authenticator for processing is `0x02`.
```

- `"userVerificationRequired"`: User verification is always required. The equivalent `credProtect` value sent to the authenticator for processing is `0x03`.

**Note:** Chromium will default to `userVerificationOptionalWithCredentialIDList` or `userVerificationRequired`, depending on the type of request:

- Chromium will request a protection level of `userVerificationOptionalWithCredentialIDList` when creating a credential if `residentKey` is set to `preferred` or `required`. (Setting `requireResidentKey` is treated the same as required.) This ensures that simple physical possession of a security key does not allow the presence of a discoverable credential for a given `rpId` to be queried.

> - Additionally, if `residentKey` is `required` and `userVerification` is preferred, the protection level will be increased to `userVerificationRequired`. This ensures that physical possession of a security key does not allow sign-in to a site that doesn't demand user verification. (This is not a complete protection; sites should still carefully consider the security of their users.)
> - If an explicit `credProtect` level is requested by the site, that will override these defaults. These defaults never cause the protection level to be lower than the security key's default, if that is higher.

Reviewer: ```suggestion
- Additionally, if `residentKey` is `required` and `userVerification` is preferred, the protection level will be increased to `userVerificationRequired`. This ensures that physical possession of a security key does not allow sign-in to a site that doesn't require user verification. (This is not complete protection; sites should still carefully consider the security of their users.)
- If the site requests an explicit `credProtect` level, that will override these defaults. These defaults never cause the protection level to be lower than the security key's default if that is higher.
```

> If the `enforceCredentialProtectionPolicy` value is `true`, the `create()` call will fail if the policy cannot be adhered to (for example, it requires user verification, but the authenticator does not support user verification). If it is `false`, the system will make a best attempt to create a credential that conforms to the policy, but it will still create one that conforms as closely as it can, if this is not possible.

Reviewer: ```suggestion
Suppose the `enforceCredentialProtectionPolicy` value is `true`. In that case, the `create()` call will fail if the policy cannot be adhered to (for example, it requires user verification, but the authenticator does not support user verification). If it is `false`, the system will make the best attempt to create a credential that conforms to the policy, but it will still create one that conforms as closely as it can if this is not possible.
```

#### Output

> If the `create()` call is successfull, the authenticator data will contain a representation of the `credProtect` value representing the set policy in the following form:

Reviewer: ```suggestion
If the `create()` call is successful, the authenticator data will contain a representation of the `credProtect` value representing the set policy in the following form:
```

```
{ "credProtect": 0x01 }
```

### `largeBlob`

- Usable in: Registration ({{domxref("CredentialsContainer.create()","create()")}}) and authentication ({{domxref("CredentialsContainer.get()","get()")}})
- Processed by: User agent
- Specification: [Large blob storage extension (largeBlob)](https://w3c.github.io/webauthn/#sctn-large-blob-extension)

> Allows a Relying Party to store blobs associated with a credential on the authenticator — for example it may wish to directly store certificates rather than run a centralised authentication service.

Reviewer: ```suggestion
Allows a relying party to store blobs associated with a credential on the authenticator — for example, it may wish to directly store certificates rather than run a centralized authentication service.
```

#### Input

During a `create()` call, the `publicKey`'s `extensions` property must contain a `largeBlob` property with the following object structure:

```js
extensions: {
  largeBlob: {
    support: "required";
  }
}
```

> The `support` property's value is a string, which can be one of the following:
> 
> - `"preferred"`: The credential will be created with an authenticator that can store blobs if possible, but it will still create one if not. The authenticator's ability to store blobs is reported by the output `supported` property.
> - `"required"`: The credential will be created with an authenticator that can store blobs. If this is not possible, the `create()` call will fail.
> 
> During a `get()` call, the `publicKey`'s `extensions` property must contain a `largeBlob` property with one of two sub-properties — `read` or `write` (`get()` fails if both are present):
> 
> The `read` property is a boolean. A value of `true` indicates that the relying party would like to fetch a previously-written blob associated with the asserted credential:

Reviewer: ```suggestion
The `support` property's value is a string, which can be one of the following:

- `"preferred"`: The credential will be created with an authenticator that can store blobs if possible, but it will still create one if not. The output' supported' property reports the authenticator's ability to store blobs.
- `"required"`: The credential will be created with an authenticator to store blobs. The `create()` call will fail if this is impossible.

During a `get()` call, the `publicKey`'s `extensions` property must contain a `largeBlob` property with one of two sub-properties — `read` or `write` (`get()` fails if both are present):

The `read` property is a boolean. A value of `true` indicates that the relying party would like to fetch a previously-written blob associated with the asserted credential:
```
